mcs/class/corlib/System/Enum.cs

```diff
@@ -402,59 +402,72 @@ public static string[] GetNames (Type enumType)
 				return Array.BinarySearch (int_array, (int)value, MonoEnumInfo.int_comparer);
+
+			uint[] uint_array = values as uint [];
+			if (uint_array != null && value is uint)
+				return Array.BinarySearch (uint_array, (uint)value);
 			
+			short [] short_array = values as short [];
+			if (short_array != null && value is short)
+				return Array.BinarySearch (short_array, (short)value, MonoEnumInfo.short_comparer);
+
 			ushort [] ushort_array = values as ushort [];
-			if (ushort_array != null)
+			if (ushort_array != null && value is ushort)
 				return Array.BinarySearch (ushort_array, (ushort)value);
 
-			short [] short_array = values as short [];
-			if (short_array != null)
-				return Array.BinarySearch (short_array, (short)value, MonoEnumInfo.short_comparer);
+			sbyte [] sbyte_array = values as sbyte [];
+			if (sbyte_array != null && value is sbyte)
+				return Array.BinarySearch (sbyte_array, (sbyte) value,  MonoEnumInfo.sbyte_comparer);
 
 			byte [] byte_array = values as byte [];
-			if (byte_array != null)
+			if (byte_array != null && value is byte)
 				return Array.BinarySearch (byte_array, (byte) value);
 
-			sbyte [] sbyte_array = values as sbyte [];
-			if (sbyte_array != null)
-				return Array.BinarySearch (sbyte_array, (sbyte) value,  MonoEnumInfo.sbyte_comparer);
+			long [] long_array = values as long [];
+			if (long_array != null && value is long)
+				return Array.BinarySearch (long_array, (long) value,  MonoEnumInfo.long_comparer);
 
 			ulong [] ulong_array = values as ulong [];
-			if (ulong_array != null)
+			if (ulong_array != null && value is ulong)
 				return Array.BinarySearch (ulong_array, (ulong) value);
-			
-			long [] long_array = values as long [];
-			if (long_array != null)
-				return Array.BinarySearch (long_array, (long) value,  MonoEnumInfo.long_comparer);
 
 			// This should never happen
 			return Array.BinarySearch (values, value);
 		}
 #else
+		*/
+
+		// After a number of attempts to workaround a bug in the implementation of isinst
+		// (see the commit at acca62e1c3b5f1bff27f637e693f57a5417bf57d for details) and subsequent
+		// problems with the FindPosition implementation above, I've decided to try this
+		// implementation, which is liekly slower, but hopefully will behave correctly both with
+		// the broken isinst behavior and with the correct isinst behavior (in il2cpp).
 		static int FindPosition (object value, Array values)
 		{
 			IComparer ic = null;
 
-			if (values is int[])
-				return Array.BinarySearch (values, value, MonoEnumInfo.int_comparer);
-			if (values is short[])
-				return Array.BinarySearch (values, value, MonoEnumInfo.short_comparer);
-			if (values is sbyte [])
-				return Array.BinarySearch (values, value,  MonoEnumInfo.sbyte_comparer);
-			if (values is long [])
-				return Array.BinarySearch (values, value,  MonoEnumInfo.long_comparer);
+			if (!(values is byte[]) &&
+				!(values is ushort[]) &&
+				!(values is uint[]) &&
+				!(values is ulong[]))
```

Comment:
This if check actually corrects the problem, by forcing us back to Array.BinarySearch without a specific comparer.